### ISSUE TITLE
Can have `Ppx_deriving_json_runtime.Primitives` working internally?

### ISSUE THREAD

**Issue by pedrobslisboa** [melange-community/jsonkit#19]
## Description

I'm curious about it. We have to add `Ppx_deriving_json_runtime.Primitives` for every usage; can we have it internally?

Like (Code just as a sample, not tested or something): 
```ocmal
let is_primitive = (* ... *)
let prefix = if (is_primitive) then ("Ppx_deriving_json_runtime.Primitives." ^ prefix) else prefix
let gen_name n = sprintf "%s_%s" prefix n in
```

Is it to avoid depending on `Ppx_deriving_json_runtime.Primitives` name, as it can break if the name changes?
What is the motivation?

**Comment by davesnx** [comment]
The only reason I can think of not doing this is to override the primitives such as string/int/..., so I propose to make things rare a bit harder for those use-cases and provide a section on the docs explaining how to do it:

```
module Ppx_deriving_json_runtime = struct
    include Ppx_deriving_json_runtime
    module Primitives = struct
        include Ppx_deriving_json_runtime.Primitives
        let string = (* ... *)
    end
end
```

and keep the inlining inside the ppx, as @pedrobslisboa suggests.

**Comment by andreypopp** [comment]
I don't think this is a good idea for the reasons stated in the linked ppx_deriving_router issue. Better to stay consistent. If the local open in a module is too annoying, or needs to be spread in multiple modules, can always add `-open ...` arguments via dune's `(flags ...)` stanza.

**Comment by davesnx** [comment]
`add -open ... arguments via dune's (flags ...) stanza.` creates all sort of issues that override to all files in the library, which is worst and I won't consider a valid solution here.

Also, I would do this change on ppx_deriving_router as well.

The entire point of this issue is to make default behaviour simpler, while allow more advanced use-cases more explicit/verbose.

- The type error that you see when you don't "open" is bad
- A ppx relying on an open module is a bad practice?

**Comment by andreypopp** [comment]
> add -open ... arguments via dune's (flags ...) stanza. creates all sort of issues that override to all files in the library, which is worst and I won't consider a valid solution here.

There's a possibility to do `open ...` per module.

> A ppx relying on an open module is a bad practice?

I disagree, all JST ppx work that way.

> The entire point of this issue is to make default behaviour simpler, while allow more advanced use-cases more explicit/verbose.

I'm all for making things simpler but what you propose will make things a bit easier, not simpler. Because special casing few types is actually making things more complex and less consistent.

**Comment by davesnx** [comment]
But special casing is shadowing type string, which is... not very common?

**Comment by davesnx** [comment]
I saw the first complain in discord:  https://discord.com/channels/235176658175262720/235176658175262720/1331705762753413132

> I disagree, all JST ppx work that way.

Looking at how jst organises code or how they do things isn't the best guide to see what's ideal.

**Comment by davesnx** [comment]
and again, making it more confusing for the 99% of users, in the tradeoff of the 1% of cases to have it more complex and less consistent is a choice I would do anytime

**Comment by pedrobslisboa** [comment]
I agree with @davesnx; it makes more sense to someone who wants to explicitly change the primitives to do it. 
I don't think having one config for that specific user can be worse than all others to apply `open` on every module.
By the way, using `open` performs worse than just using the module directly on the Dune. So, using the primitives directly to determine what is needed can probably be more performatic. (ref: https://github.com/ocaml/ocaml/issues/5877)

**Comment by pedrobslisboa** [comment]
Closing this, since we've already discussed it a lot;
What we have now: Jsonkit `json` and `jsonschema` work like this to support multiple ways to handle primitives easily. E.g., JsonSchema handles int64 differently when used for `ppx_yojson_conv` and `ppx_deriving_yojson`.
If anyone has anything more to add, we can reopen the issue.
